modopt/tests/test_signal.py

```diff
@@ -1,322 +1,233 @@
         npt.assert_raises(ValueError, svd.find_n_pc, np.arange(3))
 
-    def test_calculate_svd(self):
+    def test_calculate_svd(self, data, svd0):
         """Test calculate_svd."""
+        errors = []
+        for i, name in enumerate("USV"):
+            try:
+                npt.assert_almost_equal(svd0[i], data[2][i])
+            except AssertionError:
+                errors.append(name)
+        if errors:
+            raise AssertionError("Incorrect SVD calculation for: " + ", ".join(errors))
+
+    @pytest.mark.parametrize(
+        ("n_pc", "idx_res"), [(None, 3), (1, 4), ("all", 0), failparam("fail", 1)]
+    )
+    def test_svd_thresh(self, data, n_pc, idx_res):
+        """Test svd_tresh."""
         npt.assert_almost_equal(
-            self.svd[0],
-            np.array(self.data3)[0],
-            err_msg='Incorrect SVD calculation: U',
-        )
-
-        npt.assert_almost_equal(
-            self.svd[1],
-            np.array(self.data3)[1],
-            err_msg='Incorrect SVD calculation: S',
-        )
-
-        npt.assert_almost_equal(
-            self.svd[2],
-            np.array(self.data3)[2],
-            err_msg='Incorrect SVD calculation: V',
-        )
-
-    def test_svd_thresh(self):
-        """Test svd_thresh."""
-        npt.assert_almost_equal(
-            svd.svd_thresh(self.data1),
-            self.data4,
-            err_msg='Incorrect SVD tresholding',
-        )
-
-        npt.assert_almost_equal(
-            svd.svd_thresh(self.data1, n_pc=1),
-            self.data5,
-            err_msg='Incorrect SVD tresholding',
-        )
-
-        npt.assert_almost_equal(
-            svd.svd_thresh(self.data1, n_pc='all'),
-            self.data1,
-            err_msg='Incorrect SVD tresholding',
+            svd.svd_thresh(data[0], n_pc=n_pc),
+            data[idx_res],
         )
 
+    def test_svd_tresh_invalid_type(self):
+        """Test svd_tresh failure."""
         npt.assert_raises(TypeError, svd.svd_thresh, 1)
 
-        npt.assert_raises(ValueError, svd.svd_thresh, self.data1, n_pc='bla')
-
-    def test_svd_thresh_coef(self):
-        """Test svd_thresh_coef."""
+    @pytest.mark.parametrize("operator", [lambda x: x, failparam(0, raises=TypeError)])
+    def test_svd_thresh_coef(self, data, operator):
+        """Test svd_tresh_coef."""
         npt.assert_almost_equal(
-            svd.svd_thresh_coef(self.data1, lambda x_val: x_val, 0),
-            self.data1,
-            err_msg='Incorrect SVD coefficient tresholding',
+            svd.svd_thresh_coef(data[0], operator, 0),
+            data[0],
+            err_msg="Incorrect SVD coefficient tresholding",
         )
 
-        npt.assert_raises(TypeError, svd.svd_thresh_coef, self.data1, 0, 0)
+    # TODO test_svd_thresh_coef_fast
 
 
-class ValidationTestCase(TestCase):
-    """Test case for validation module."""
+# TODO: is this module really necessary ?
+# It is not use anywhere, neither in  modopt, nor pysap.
```

Comment:
This can be quite a useful tool for signal processing tests, so I don't see any reason to remove it (for now at least).

Reply:
OK. lets keep it then :)